syntax = "proto3";

package fx.descriptor.v1beta;

// Workspace -------------------------------------------------------------------

message FxWorkspaceDescriptor {
    string descriptor_version = 1;
    repeated string ignore = 2;
}

// Command ---------------------------------------------------------------------

message FxCommandDescriptor {
    string descriptor_version = 1;
    string synopsis = 2;
    optional string description = 3;
    repeated OptionDescriptor options = 4;
    repeated ArgumentDescriptor arguments = 5;
    RuntimeDescriptor runtime = 6;
}

message OptionDescriptor {
    string name = 1;
    optional string short_name = 2;
    string description = 3;
    oneof value {
        BoolValueDescriptor bool_value = 100;
        IntValueDescriptor int_value = 101;
        DoubleValueDescriptor double_value = 102;
        StringValueDescriptor string_value = 103;
    }
}

message ArgumentDescriptor {
    string name = 1;
    string description = 2;
    oneof value {
        IntValueDescriptor int_value = 100;
        DoubleValueDescriptor double_value = 101;
        StringValueDescriptor string_value = 102;
    }
}

message BoolValueDescriptor {}

message StringValueDescriptor {
    repeated string choices = 1;
    string default = 2;
    bool required = 3;
    bool list = 4;
}

message IntValueDescriptor {
    repeated int64 choices = 1;
    int64 default = 2;
    bool required = 3;
    bool list = 4;
}

message DoubleValueDescriptor {
    repeated double choices = 1;
    double default = 2;
    bool required = 3;
    bool list = 4;
}

message RuntimeDescriptor {
    string run = 1;
}
